File: Assignment 3 /task 2.2 and task 2.3/task2_2.py
import sys
import time
import json
import math
import numpy as np
from xgboost import XGBRegressor
from pyspark import SparkContext, SparkConf

# Handle collections warning in newer Python versions
import collections
import collections.abc
collections.Mapping = collections.abc.Mapping

# ===================================================================
# DATA LOADING FUNCTIONS
# ===================================================================

def read_csv_data(file_path, include_ratings=True):
    """
    Reads CSV data with appropriate error handling.
    
    Args:
        file_path (str): Path to the CSV file
        include_ratings (bool): Whether to include ratings column
        
    Returns:
        RDD of parsed rows
    """
    # Load the raw CSV file as an RDD of text lines
    raw_data = sc.textFile(file_path)
    
    # Extract the header line
    header = raw_data.first()
    
    # Split rows and filter out malformed ones
    parsed_data = (raw_data.filter(lambda row: row != header)
                   .map(lambda row: row.split(",")))
    
    # Map to appropriate format based on whether we need ratings
    if include_ratings:
        return parsed_data.filter(lambda r: len(r) >= 3)
    else:
        return parsed_data.filter(lambda r: len(r) >= 2)

def load_json_data(file_path, mapper_function):
    """
    Loads and processes JSON data.
    
    Args:
        file_path (str): Path to the JSON file
        mapper_function (function): Function to map each JSON object
        
    Returns:
        RDD of mapped data
    """
    # Load file and parse JSON
    data = sc.textFile(file_path)
    return data.map(lambda row: json.loads(row)).map(mapper_function)

# ===================================================================
# FEATURE EXTRACTION FUNCTIONS
# ===================================================================

def extract_review_features(folder_path):
    """
    Extracts features from review data.
    
    Args:
        folder_path (str): Path to the folder containing review data
        
    Returns:
        dict: Mapping of business_id to review features (useful, funny, cool)
    """
    # ---------------------------
    # 1. Load review data
    # ---------------------------
    review_mapper = lambda row: (row['business_id'], 
                               (float(row['useful']), float(row['funny']), float(row['cool'])))
    review_data = load_json_data(folder_path + '/review_train.json', review_mapper)
    
    # Group by business_id
    review_grouped = review_data.groupByKey().mapValues(list)
    
    # ---------------------------
    # 2. Collect and process review data
    # ---------------------------
    # First collect the raw review data
    review_dict = {}
    for business_id, features_list in review_grouped.collect():
        review_dict[business_id] = features_list
    
    # Calculate average metrics per business
    business_review_features = {}
    for business_id, reviews in review_dict.items():
        if not reviews:
            continue
            
        useful_sum = funny_sum = cool_sum = 0
        count = len(reviews)
        
        for useful, funny, cool in reviews:
            useful_sum += useful
            funny_sum += funny
            cool_sum += cool
            
        business_review_features[business_id] = (
            useful_sum / count,  # Average useful rating
            funny_sum / count,   # Average funny rating
            cool_sum / count     # Average cool rating
        )
        
    return business_review_features

def extract_user_features(folder_path):
    """
    Extracts features from user data.
    
    Args:
        folder_path (str): Path to the folder containing user data
        
    Returns:
        dict: Mapping of user_id to user features (avg_stars, review_count, fans)
    """
    # ---------------------------
    # 1. Load user data
    # ---------------------------
    user_mapper = lambda row: (row['user_id'], 
                             (float(row['average_stars']), 
                              float(row['review_count']), 
                              float(row['fans'])))
    user_data = load_json_data(folder_path + '/user.json', user_mapper)
    
    # ---------------------------
    # 2. Collect user data
    # ---------------------------
    user_features = {}
    for user_id, features in user_data.collect():
        user_features[user_id] = features
        
    return user_features

def extract_business_features(folder_path):
    """
    Extracts features from business data.
    
    Args:
        folder_path (str): Path to the folder containing business data
        
    Returns:
        dict: Mapping of business_id to business features (stars, review_count)
    """
    # ---------------------------
    # 1. Load business data
    # ---------------------------
    business_mapper = lambda row: (row['business_id'], 
                                 (float(row['stars']), 
                                  float(row['review_count'])))
    business_data = load_json_data(folder_path + '/business.json', business_mapper)
    
    # ---------------------------
    # 2. Collect business data
    # ---------------------------
    business_features = {}
    for business_id, features in business_data.collect():
        business_features[business_id] = features
        
    return business_features

# ===================================================================
# FEATURE VECTOR BUILDING FUNCTIONS
# ===================================================================

def build_feature_vector(user_id, business_id, business_review_features, user_features, business_features):
    """
    Builds a feature vector for a user-business pair.
    
    Args:
        user_id (str): ID of the user
        business_id (str): ID of the business
        business_review_features (dict): Review features per business
        user_features (dict): Features per user
        business_features (dict): Features per business
        
    Returns:
        list: Feature vector
    """
    # ---------------------------
    # 1. Extract review features
    # ---------------------------
    if business_id in business_review_features:
        useful, funny, cool = business_review_features[business_id]
    else:
        # Default values if missing
        useful = funny = cool = 0.0
    
    # ---------------------------
    # 2. Extract user features
    # ---------------------------
    if user_id in user_features:
        user_avg_stars, user_review_count, user_fans = user_features[user_id]
    else:
        # Default values if missing
        user_avg_stars = 3.5  # Global average
        user_review_count = 0.0
        user_fans = 0.0
    
    # ---------------------------
    # 3. Extract business features
    # ---------------------------
    if business_id in business_features:
        business_avg_stars, business_review_count = business_features[business_id]
    else:
        # Default values if missing
        business_avg_stars = 3.5  # Global average
        business_review_count = 0.0
    
    # ---------------------------
    # 4. Build complete feature vector
    # ---------------------------
    # Combine all features into a single vector
    return [
        useful, funny, cool,
        user_avg_stars, user_review_count, user_fans,
        business_avg_stars, business_review_count
    ]

def prepare_training_data(train_data, business_review_features, user_features, business_features):
    """
    Prepares feature vectors and target values for model training.
    
    Args:
        train_data (RDD): Training data with user_id, business_id, rating
        business_review_features (dict): Review features per business
        user_features (dict): Features per user
        business_features (dict): Features per business
        
    Returns:
        tuple: (X_train, y_train) as numpy arrays
    """
    X_train = []  # Feature vectors
    y_train = []  # Target ratings
    
    # For each training sample, create a feature vector
    for row in train_data.collect():
        user_id, business_id, rating = row[0], row[1], float(row[2])
        
        # Store target rating
        y_train.append(rating)
        
        # Build feature vector
        feature_vector = build_feature_vector(
            user_id, business_id, 
            business_review_features, user_features, business_features
        )
        
        X_train.append(feature_vector)
    
    # Convert to numpy arrays for XGBoost
    return np.array(X_train, dtype='float32'), np.array(y_train, dtype='float32')

def prepare_test_data(test_data, business_review_features, user_features, business_features):
    """
    Prepares feature vectors for prediction.
    
    Args:
        test_data (RDD): Test data with user_id, business_id
        business_review_features (dict): Review features per business
        user_features (dict): Features per user
        business_features (dict): Features per business
        
    Returns:
        tuple: (user_business_pairs, X_test)
    """
    user_business_pairs = []
    X_test = []
    
    # For each test sample, create a feature vector
    for row in test_data.collect():
        user_id, business_id = row[0], row[1]
        
        # Store user-business pair for output
        user_business_pairs.append((user_id, business_id))
        
        # Build feature vector
        feature_vector = build_feature_vector(
            user_id, business_id, 
            business_review_features, user_features, business_features
        )
        
        X_test.append(feature_vector)
    
    # Convert to numpy array for XGBoost
    return user_business_pairs, np.array(X_test, dtype='float32')

# ===================================================================
# MODEL TRAINING AND PREDICTION
# ===================================================================

def train_xgboost_model(X_train, y_train):
    """
    Trains an XGBoost regression model.
    
    Args:
        X_train (numpy.array): Training feature vectors
        y_train (numpy.array): Target rating values
        
    Returns:
        XGBRegressor: Trained model
    """
    # ---------------------------
    # 1. Set model parameters
    # ---------------------------
    # These parameters are tuned for good performance
    params = {
        'learning_rate': 0.02,
        'max_depth': 17,
        'subsample': 0.8,
        'colsample_bytree': 0.5,
        'min_child_weight': 101,
        'n_estimators': 300,
        'random_state': 42
    }
    
    # ---------------------------
    # 2. Initialize and train model
    # ---------------------------
    model = XGBRegressor(**params)
    model.fit(X_train, y_train)
    
    return model

def predict_ratings(model, X_test):
    """
    Generates predictions from a trained model.
    
    Args:
        model (XGBRegressor): Trained model
        X_test (numpy.array): Test feature vectors
        
    Returns:
        numpy.array: Predicted ratings
    """
    # Generate raw predictions
    predictions = model.predict(X_test)
    
    # Ensure predictions are within valid range [1.0, 5.0]
    return np.clip(predictions, 1.0, 5.0)

# ===================================================================
# EVALUATION FUNCTIONS
# ===================================================================

def calculate_rmse(predictions, actual_ratings):
    """
    Calculates Root Mean Square Error between predictions and actual ratings.
    
    Args:
        predictions (list): List of predicted ratings
        actual_ratings (list): List of actual ratings
        
    Returns:
        float: RMSE value
    """
    if len(predictions) != len(actual_ratings):
        raise ValueError("Prediction and actual rating lists must be the same length")
        
    squared_errors = [(pred - actual) ** 2 for pred, actual in zip(predictions, actual_ratings)]
    mean_squared_error = sum(squared_errors) / len(squared_errors)
    
    return math.sqrt(mean_squared_error)

def calculate_error_distribution(predictions, actual_ratings):
    """
    Calculates the distribution of prediction errors across different ranges.
    
    Args:
        predictions (list): List of predicted ratings
        actual_ratings (list): List of actual ratings
        
    Returns:
        dict: Dictionary with error ranges as keys and counts as values
    """
    error_ranges = {
        ">=0 and <1": 0,
        ">=1 and <2": 0,
        ">=2 and <3": 0,
        ">=3 and <4": 0,
        ">=4": 0
    }
    
    for pred, actual in zip(predictions, actual_ratings):
        error = abs(pred - actual)
        
        if error < 1:
            error_ranges[">=0 and <1"] += 1
        elif error < 2:
            error_ranges[">=1 and <2"] += 1
        elif error < 3:
            error_ranges[">=2 and <3"] += 1
        elif error < 4:
            error_ranges[">=3 and <4"] += 1
        else:
            error_ranges[">=4"] += 1
            
    return error_ranges

# ===================================================================
# MAIN FUNCTION
# ===================================================================

def main():
    """
    Main function for Task 2.2: Model-based Recommendation System
    """
    # ---------------------------
    # 1. Setup and Initialization
    # ---------------------------
    # Start timing execution
    start_time = time.time()
    
    # Parse command line arguments
    if len(sys.argv) != 4:
        print("Usage: spark-submit task2_2.py <folder_path> <test_file_name> <output_file_name>")
        sys.exit(1)
        
    folder_path = sys.argv[1]
    test_file = sys.argv[2]
    output_file = sys.argv[3]
    
    # Initialize Spark context
    conf = SparkConf().setAppName("Task2_2_ModelBasedRS")
    global sc
    sc = SparkContext(conf=conf)
    sc.setLogLevel("ERROR")  # Reduce log verbosity
    
    # ---------------------------
    # 2. Extract Features
    # ---------------------------
    print("Extracting features...")
    
    # Extract review features
    business_review_features = extract_review_features(folder_path)
    
    # Extract user features
    user_features = extract_user_features(folder_path)
    
    # Extract business features
    business_features = extract_business_features(folder_path)
    
    # ---------------------------
    # 3. Prepare Training Data
    # ---------------------------
    print("Preparing training data...")
    
    # Load training data
    train_data = read_csv_data(folder_path + '/yelp_train.csv')
    
    # Create feature vectors and target values
    X_train, y_train = prepare_training_data(
        train_data, business_review_features, user_features, business_features
    )
    
    # ---------------------------
    # 4. Train Model
    # ---------------------------
    print("Training XGBoost model...")
    
    # Train XGBoost model
    model = train_xgboost_model(X_train, y_train)
    
    # ---------------------------
    # 5. Prepare Test Data
    # ---------------------------
    print("Preparing test data...")
    
    # Load test data
    test_data = read_csv_data(test_file, include_ratings=False)
    
    # Create feature vectors for prediction
    user_business_pairs, X_test = prepare_test_data(
        test_data, business_review_features, user_features, business_features
    )
    
    # ---------------------------
    # 6. Generate Predictions
    # ---------------------------
    print("Generating predictions...")
    
    # Make predictions using trained model
    predictions = predict_ratings(model, X_test)
    
    # ---------------------------
    # 7. Write Output File
    # ---------------------------
    print("Writing output file...")
    
    # Write predictions to CSV
    with open(output_file, "w") as f:
        f.write("user_id,business_id,prediction\n")
        for i, (user_id, business_id) in enumerate(user_business_pairs):
            f.write(f"{user_id},{business_id},{predictions[i]}\n")
    
    # ---------------------------
    # 8. Calculate RMSE (if test file has ratings)
    # ---------------------------
    try:
        # Try to read test data with ratings
        test_data_with_ratings = read_csv_data(test_file, include_ratings=True)
        test_ratings = test_data_with_ratings.map(lambda r: (r[0], r[1], float(r[2]))).collect()
        
        # Create dictionaries for lookup
        actual_ratings_dict = {(user_id, business_id): rating for user_id, business_id, rating in test_ratings}
        actual_ratings = []
        
        # Extract actual ratings in the same order as predictions
        for i, (user_id, business_id) in enumerate(user_business_pairs):
            if (user_id, business_id) in actual_ratings_dict:
                actual_ratings.append(actual_ratings_dict[(user_id, business_id)])
            else:
                # If no actual rating, remove the corresponding prediction
                predictions[i] = None
        
        # Remove None values
        filtered_predictions = [p for p in predictions if p is not None]
        
        # Calculate and print RMSE
        if filtered_predictions:
            rmse = calculate_rmse(filtered_predictions, actual_ratings)
            print(f"RMSE: {rmse}")
            
            # Calculate error distribution
            error_dist = calculate_error_distribution(filtered_predictions, actual_ratings)
            print("Error Distribution:")
            for range_label, count in error_dist.items():
                print(f"  {range_label}: {count}")
    except:
        # If RMSE calculation fails, just skip it
        pass
    
    # ---------------------------
    # 9. Finalization
    # ---------------------------
    # Print execution time
    execution_time = time.time() - start_time
    print(f"Task 2.2 completed successfully in {execution_time:.2f} seconds.")
    
    # Stop Spark context
    sc.stop()

if __name__ == "__main__":
    main()

File: Assignment 3 /task 2.2 and task 2.3/task2_2.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.rdd.RDD

import org.json4s._
import org.json4s.jackson.JsonMethods._

import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.mllib.linalg.Vectors

// Spark MLlib for Gradient-Boosted Trees
import org.apache.spark.mllib.tree.GradientBoostedTrees
import org.apache.spark.mllib.tree.configuration.BoostingStrategy

import java.io.{File, PrintWriter}
import scala.math._

object task2_2 {

  // -------------------------------------------------------------------
  // Main entry point
  // -------------------------------------------------------------------
  def main(args: Array[String]): Unit = {

    // -------------------------------
    // 1. Parse arguments
    // -------------------------------
    if (args.length != 3) {
      println("Usage: spark-submit --class hw3.task2_2 hw3.jar <folder_path> <test_file_name> <output_file_name>")
      sys.exit(1)
    }
    val folderPath = args(0)      // Path to folder containing the JSON files + yelp_train.csv
    val testFile   = args(1)      // Path to test CSV file
    val outputFile = args(2)      // Where we write predictions

    // -------------------------------
    // 2. Spark Context
    // -------------------------------
    val conf = new SparkConf().setAppName("Task2_2_ModelBasedRS")
    val sc = new SparkContext(conf)
    sc.setLogLevel("ERROR")

    val startTime = System.currentTimeMillis()

    // -------------------------------
    // 3. Load & Process JSON Data
    // -------------------------------
    // We will parse:
    //   - review_train.json -> average (useful, funny, cool) per business
    //   - user.json -> (average_stars, review_count, fans) per user
    //   - business.json -> (stars, review_count) per business
    //
    // Then collect them into driver maps for fast lookups (broadcast later).

    // a) REVIEW data
    val reviewPath   = s"$folderPath/review_train.json"
    val reviewLines  = sc.textFile(reviewPath)

    // Parse each line as JSON. We do a manual parse with json4s.
    val businessReviewRDD: RDD[(String, (Double, Double, Double))] = reviewLines.mapPartitions { iter =>
      implicit val formats: DefaultFormats.type = DefaultFormats
      iter.map { line =>
        val jval = parse(line)
        val businessId = (jval \ "business_id").extract[String]
        val useful  = (jval \ "useful").extract[Double]
        val funny   = (jval \ "funny").extract[Double]
        val cool    = (jval \ "cool").extract[Double]
        (businessId, (useful, funny, cool))
      }
    }

    // Compute average (useful, funny, cool) for each business
    val reviewFeatures = businessReviewRDD
      .groupByKey()
      .mapValues { feats =>
        val n = feats.size
        val sums = feats.foldLeft((0.0, 0.0, 0.0)) {
          case ((su, sf, sc), (u, f, c)) => (su + u, sf + f, sc + c)
        }
        (sums._1 / n, sums._2 / n, sums._3 / n)
      }
      .collectAsMap()

    // b) USER data
    val userPath  = s"$folderPath/user.json"
    val userLines = sc.textFile(userPath)

    val userRDD: RDD[(String, (Double, Double, Double))] = userLines.mapPartitions { iter =>
      implicit val formats: DefaultFormats.type = DefaultFormats
      iter.map { line =>
        val jval = parse(line)
        val user_id       = (jval \ "user_id").extract[String]
        val average_stars = (jval \ "average_stars").extract[Double]
        val review_count  = (jval \ "review_count").extract[Double]
        val fans          = (jval \ "fans").extract[Double]
        (user_id, (average_stars, review_count, fans))
      }
    }
    val userFeatures = userRDD.collectAsMap()

    // c) BUSINESS data
    val busPath  = s"$folderPath/business.json"
    val busLines = sc.textFile(busPath)

    val busRDD: RDD[(String, (Double, Double))] = busLines.mapPartitions { iter =>
      implicit val formats: DefaultFormats.type = DefaultFormats
      iter.map { line =>
        val jval = parse(line)
        val businessId   = (jval \ "business_id").extract[String]
        val stars        = (jval \ "stars").extract[Double]
        val review_count = (jval \ "review_count").extract[Double]
        (businessId, (stars, review_count))
      }
    }
    val businessFeatures = busRDD.collectAsMap()

    // Broadcast all maps
    val brReviewMap   = sc.broadcast(reviewFeatures)
    val brUserMap     = sc.broadcast(userFeatures)
    val brBusinessMap = sc.broadcast(businessFeatures)

    // -------------------------------
    // 4. Load Training Data (yelp_train.csv) & Prepare Features
    // -------------------------------
    val trainFile = s"$folderPath/yelp_train.csv"
    val trainRaw  = sc.textFile(trainFile)
    val trainHeader = trainRaw.first()

    // (user_id, business_id, rating)
    val trainRDD = trainRaw
      .filter(_ != trainHeader)
      .map(_.split(","))
      .filter(_.length >= 3)
      .map { arr =>
        val userId = arr(0)
        val busId  = arr(1)
        val rating = arr(2).toDouble
        (userId, busId, rating)
      }

    // Build LabeledPoint for each record:
    //  [ avg_useful, avg_funny, avg_cool,
    //    user_avg_stars, user_review_count, user_fans,
    //    business_avg_stars, business_review_count ]
    val trainingData = trainRDD.map { case (userId, busId, rating) =>
      val (useful, funny, cool) = brReviewMap.value.getOrElse(busId, (0.0, 0.0, 0.0))
      val (uStars, uReviews, uFans) = brUserMap.value.getOrElse(userId, (3.5, 0.0, 0.0))
      val (bStars, bReviews) = brBusinessMap.value.getOrElse(busId, (3.5, 0.0))

      val features = Vectors.dense(
        useful, funny, cool,
        uStars, uReviews, uFans,
        bStars, bReviews
      )
      LabeledPoint(rating, features)
    }.cache()

    // -------------------------------
    // 5. Train Model (Gradient Boosted Trees)
    // -------------------------------
    // "trainRegressor(...)" is not present in some Spark distributions. 
    // We can use "train(...)" with a 'Regression' boostingStrategy.
    val boostingStrategy = BoostingStrategy.defaultParams("Regression")
    boostingStrategy.numIterations = 55      // # of trees
    boostingStrategy.treeStrategy.maxDepth = 7

    // Train the model
    val gbtModel = GradientBoostedTrees.train(trainingData, boostingStrategy)

    // -------------------------------
    // 6. Prepare Test Data
    // -------------------------------
    val testRaw   = sc.textFile(testFile)
    val testHeader= testRaw.first()
    val testRDD   = testRaw
      .filter(_ != testHeader)
      .map(_.split(","))
      .filter(_.length >= 2)
      .map(arr => (arr(0), arr(1)))  // (user_id, business_id)

    // We'll collect so we can write in the same order. 
    // For very large data, you might do a distributed write.
    val testPairs = testRDD.collect()

    // -------------------------------
    // 7. Predict
    // -------------------------------
    val broadcastPairs = sc.broadcast(testPairs)
    val predictions: Array[(String, String, Double)] = broadcastPairs.value.map { case (userId, busId) =>
      val (useful, funny, cool)     = brReviewMap.value.getOrElse(busId, (0.0, 0.0, 0.0))
      val (uStars, uReviews, uFans) = brUserMap.value.getOrElse(userId, (3.5, 0.0, 0.0))
      val (bStars, bReviews)        = brBusinessMap.value.getOrElse(busId, (3.5, 0.0))

      val featVec = Vectors.dense(useful, funny, cool, uStars, uReviews, uFans, bStars, bReviews)
      val rawPred = gbtModel.predict(featVec)

      // Clip to [1.0, 5.0]
      val finalPred = math.max(1.0, math.min(5.0, rawPred))
      (userId, busId, finalPred)
    }

    // -------------------------------
    // 8. Write Output CSV
    // -------------------------------
    val writer = new PrintWriter(new File(outputFile))
    try {
      writer.write("user_id,business_id,prediction\n")
      predictions.foreach { case (u, b, p) =>
        writer.write(s"$u,$b,$p\n")
      }
    } finally {
      writer.close()
    }

    // -------------------------------
    // 9. [Optional] RMSE if test file has ratings
    // -------------------------------
    try {
      val testDataWithRatings = testRaw
        .filter(_ != testHeader)
        .map(_.split(","))
        .filter(_.length >= 3)
        .map(arr => ((arr(0), arr(1)), arr(2).toDouble))

      val actualMap = testDataWithRatings.collectAsMap()
      var seSum = 0.0
      var count = 0L

      predictions.foreach { case (u, b, pred) =>
        val key = (u, b)
        if (actualMap.contains(key)) {
          val actual = actualMap(key)
          val err = pred - actual
          seSum += (err * err)
          count += 1
        }
      }

      if (count > 0) {
        val rmse = math.sqrt(seSum / count)
        println(f"RMSE on provided test file: $rmse%.4f (based on $count%d labeled records)")
      }
    } catch {
      case _: Throwable => // Test file might not have ratings
    }

    // -------------------------------
    // 10. Done
    // -------------------------------
    val duration = (System.currentTimeMillis() - startTime) / 1000.0
    println(f"Task2_2 completed in $duration%.2f seconds.")
    sc.stop()
  }
}